File: JantungIn_API/src/models/Diagnosis.js
'use strict';

const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');
const User = require('./User');

// Sequelize model definition for PostgreSQL
const Diagnosis = sequelize.define('Diagnosis', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  userId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: User,
      key: 'id',
    },
  },
  createdBy: {
    type: DataTypes.UUID,
    allowNull: true, // Bisa null untuk backward compatibility
    references: {
      model: User,
      key: 'id',
    },
  },
  age: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  sex: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  chestPainType: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  restingEcgResults: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  fastingBloodSugar: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  restingBloodPressure: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  maximumHeartRate: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  exerciseInducedAngina: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  stSegment: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  majorVessels: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  thalassemia: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  serumCholesterol: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  stDepression: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  resultPercentage: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  cardiovascularRisk: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  prediction: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: 'Berisiko', // Default value untuk backward compatibility
  },
});

// Definisi relasi
User.hasMany(Diagnosis, { foreignKey: 'userId', as: 'patientDiagnoses' });
Diagnosis.belongsTo(User, { foreignKey: 'userId', as: 'patient' });

// Relasi untuk pembuat diagnosis (dokter/admin)
User.hasMany(Diagnosis, { foreignKey: 'createdBy', as: 'createdDiagnoses' });
Diagnosis.belongsTo(User, { foreignKey: 'createdBy', as: 'creator' });

module.exports = Diagnosis;

File: JantungIn_Go_Latest/internal/data/entity/diagnosis.go
package entity

import (
	"time"

	"github.com/google/uuid"
)

type Diagnosis struct {
	ID                    uuid.UUID  `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	UserID                uuid.UUID  `gorm:"type:uuid;not null" json:"userId"`
	CreatedBy             *uuid.UUID `gorm:"type:uuid" json:"createdBy"` // Bisa null
	Age                   int        `gorm:"not null" json:"age"`
	Sex                   string     `gorm:"not null" json:"sex"`
	ChestPainType         string     `gorm:"not null" json:"chestPainType"`
	RestingEcgResults     string     `gorm:"not null" json:"restingEcgResults"`
	FastingBloodSugar     float64    `gorm:"not null" json:"fastingBloodSugar"`
	RestingBloodPressure  float64    `gorm:"not null" json:"restingBloodPressure"`
	MaximumHeartRate      int        `gorm:"not null" json:"maximumHeartRate"`
	ExerciseInducedAngina string     `gorm:"not null" json:"exerciseInducedAngina"`
	StSegment             string     `gorm:"not null" json:"stSegment"`
	MajorVessels          int        `gorm:"not null" json:"majorVessels"`
	Thalassemia           string     `gorm:"not null" json:"thalassemia"`
	SerumCholesterol      float64    `gorm:"not null" json:"serumCholesterol"`
	StDepression          float64    `gorm:"not null" json:"stDepression"`
	ResultPercentage      float64    `gorm:"not null" json:"resultPercentage"`
	CardiovascularRisk    string     `gorm:"not null" json:"cardiovascularRisk"`
	Prediction            string     `gorm:"default:'Berisiko';not null" json:"prediction"`
	CreatedAt             time.Time  `json:"createdAt"`
	UpdatedAt             time.Time  `json:"updatedAt"`

	// Relasi
	Patient User `gorm:"foreignKey:UserID" json:"patient,omitempty"`
	Creator User `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
}
